src/query/composable_dispatcher.rs

```diff
@@ -0,0 +1,664 @@
+            inner: impl Fn(&Isometry<Real>, &'a S1, &'b S2) -> bool + 'c + Copy,
+        ) -> Box<dyn Fn(&Isometry<Real>, &'a dyn Shape, &'b dyn Shape) -> Result<bool, ()> + 'c>
+        {
+            Box::new(
+                move |pose: &Isometry<Real>, s1: &dyn Shape, s2: &dyn Shape| {
+                    to_custom(
+                        |s: &'a dyn Shape| s.as_shape::<S1>(),
+                        |s: &'b dyn Shape| s.as_shape::<S2>(),
+                        inner,
+                    )(pose, s1, s2)
+                },
+            )
+        }
+        /// Transform a function that takes two concrete shapes to a function that takes two dynamic shapes.
+        // TODO: this would be great to have as const, or a macro.
+        fn to_custom<'c, 'a: 'c, 'b: 'c, ShapeIn1: 'a, ShapeIn2: 'b>(
+            shape_1: impl Fn(&'a dyn Shape) -> Option<ShapeIn1> + 'a,
+            shape_2: impl Fn(&'b dyn Shape) -> Option<ShapeIn2> + 'b,
+            inner: impl Fn(&Isometry<Real>, ShapeIn1, ShapeIn2) -> bool + 'c,
+        ) -> Box<dyn Fn(&Isometry<Real>, &'a dyn Shape, &'b dyn Shape) -> Result<bool, ()> + 'c>
+        {
+            Box::new(
+                move |pose: &Isometry<Real>, s1: &dyn Shape, s2: &dyn Shape| {
+                    let shape1 = shape_1(s1).ok_or(())?;
+                    let shape2 = shape_2(s2).ok_or(())?;
+                    Ok(inner(pose, shape1, shape2))
+                },
+            )
+        }
+
+        let functions = vec![
+            Box::new(|pose: &Isometry<Real>, s1: &dyn Shape, s2: &dyn Shape| {
+                Ok(query::details::intersection_test_ball_ball(
+                    &pose.translation.vector.into(),
+                    s1.as_ball().ok_or(())?,
+                    s2.as_ball().ok_or(())?,
+                ))
+            }),
+            to_as_shape(query::details::intersection_test_cuboid_cuboid),
+            to_as_shape(query::details::intersection_test_cuboid_triangle),
+            to_as_shape(query::details::intersection_test_triangle_cuboid),
+            to_custom(
+                |s: &dyn Shape| s.as_shape::<Ball>(),
+                |s: &dyn Shape| Some(s),
+                query::details::intersection_test_ball_point_query,
+            ),
+            to_custom(
+                |s: &dyn Shape| Some(s),
+                |s: &dyn Shape| s.as_shape::<Ball>(),
+                query::details::intersection_test_point_query_ball,
+            ),
+            to_custom(
+                |s: &dyn Shape| s.as_shape::<HalfSpace>(),
+                |s: &dyn Shape| s.as_support_map(),
+                query::details::intersection_test_halfspace_support_map,
+            ),
+            to_custom(
+                |s: &dyn Shape| s.as_support_map(),
+                |s: &dyn Shape| s.as_shape::<HalfSpace>(),
+                query::details::intersection_test_support_map_halfspace,
+            ),
+        ];
```

Comment:
only `intersection_test` has been changed compared to default dispatcher.

To be closer to existing code, I cannot rely on `TypeId` as `as_cuboid` etc rely on trait implementation, potentially implemented for different types. So a "parameter conversion" step has to be taken into account.

It's especially more important for "as_support_map".